pkg/controller/elasticsearch/stackmon/container.go

```diff
@@ -0,0 +1,176 @@
+		}
+		builder.WithContainers(metricbeat)
+	}
+
+	if isMonitoringLogs {
+		// Enable Stack logging to write Elasticsearch logs to disk
+		builder.WithEnv(stackLoggingEnvVar())
+
+		filebeatVolumes := append(
+			monitoringLogsTargetCaCertSecretVolumes(es),
+			filebeatConfigSecretVolume(es),
+		)
+		volumeLikes = append(volumeLikes, filebeatVolumes...)
+
+		// Inject Filebeat sidecar container
+		filebeat, err := filebeatContainer(es, filebeatVolumes)
+		if err != nil {
+			return nil, err
+		}
+		builder.WithContainers(filebeat)
+	}
+
+	// Inject volumes
+	volumes := make([]corev1.Volume, 0)
+	for _, v := range volumeLikes {
+		volumes = append(volumes, v.Volume())
+	}
+	builder.WithVolumes(volumes...)
+
+	return builder, nil
+}
+
+func stackLoggingEnvVar() corev1.EnvVar {
+	return corev1.EnvVar{Name: esLogStyleEnvVarKey, Value: esLogStyleEnvVarValue}
+}
+
+func metricbeatContainer(es esv1.Elasticsearch, volumes []volume.VolumeLike) (corev1.Container, error) {
+	image, err := fullContainerImage(es, container.MetricbeatImage)
+	if err != nil {
+		return corev1.Container{}, err
+	}
+
+	volumeMounts := make([]corev1.VolumeMount, 0)
+	for _, v := range volumes {
+		volumeMounts = append(volumeMounts, v.VolumeMount())
+	}
+
+	envVars := append(monitoringSourceEnvVars(es), monitoringTargetEnvVars(es.GetMonitoringMetricsAssociation())...)
+
+	return corev1.Container{
+		Name:         metricbeatContainerName,
+		Image:        image,
+		Args:         []string{"-c", filepath.Join(metricbeatConfigDirMountPath, metricbeatConfigKey), "-e"},
+		Env:          append(envVars, defaults.PodDownwardEnvVars()...),
+		VolumeMounts: volumeMounts,
+	}, nil
+}
+
+func filebeatContainer(es esv1.Elasticsearch, volumes []volume.VolumeLike) (corev1.Container, error) {
+	image, err := fullContainerImage(es, container.FilebeatImage)
+	if err != nil {
+		return corev1.Container{}, err
+	}
+
+	volumeMounts := []corev1.VolumeMount{
+		esvolume.DefaultLogsVolumeMount, // mount Elasticsearch logs volume into the Filebeat container
+	}
+	for _, v := range volumes {
+		volumeMounts = append(volumeMounts, v.VolumeMount())
+	}
+
+	envVars := monitoringTargetEnvVars(es.GetMonitoringLogsAssociation())
+
+	return corev1.Container{
+		Name:         filebeatContainerName,
+		Image:        image,
+		Args:         []string{"-c", filepath.Join(filebeatConfigDirMountPath, filebeatConfigKey), "-e"},
+		Env:          append(envVars, defaults.PodDownwardEnvVars()...),
+		VolumeMounts: volumeMounts,
+	}, nil
+}
+
+// fullContainerImage returns the full Beat container image with the image registry.
+// If the Elasticsearch specification is configured with a custom image, we do best effort by trying to derive the Beat
+// image from the Elasticsearch custom image with an image name replacement
+// (<registry>/elasticsearch/elasticsearch:<version> becomes <registry>/beats/<filebeat|metricbeat>:<version>)
+func fullContainerImage(es esv1.Elasticsearch, defaultImage container.Image) (string, error) {
```

Comment:
I'm wondering if this is not a bit too much?
vs. clearly documenting how users can override the Beats image through the podTemplate?

Reply:
I think you are right. I prefer to document the podTemplate override than that.